app/assets/javascripts/components/autocomplete.js

```diff
@@ -4,18 +4,63 @@ var $autocompletes = document.querySelectorAll('[data-module="autocomplete"]')
 if ($autocompletes) {
   $autocompletes.forEach(function ($el) {
     var customAttributes = {}
-    if ($el.attributes['data-contextual-guidance']) {
+    var $select = $el.querySelector('select')
+    if ($select.attributes['data-contextual-guidance']) {
       customAttributes = {
-        'data-contextual-guidance': $el.attributes['data-contextual-guidance'].value
+        'data-contextual-guidance': $select.attributes['data-contextual-guidance'].value
       }
     }
 
     // disabled eslint because we can not control the name of the constructor (expected to be EnhanceSelectElement)
     new window.accessibleAutocomplete.enhanceSelectElement({ // eslint-disable-line no-new, new-cap
-      selectElement: $el,
+      selectElement: $select,
       minLength: 3,
       showNoOptionsFound: true,
       customAttributes: customAttributes
     })
   })
 }
+
+var $autocompleteWithHintOnOptions = document.querySelector('[data-module="autocomplete-with-hint-on-options"]')
+if ($autocompleteWithHintOnOptions) {
+  // Read options and associated data attributes and feed that as results for inputValueTemplate
+  var $select = $autocompleteWithHintOnOptions.querySelector('select')
+  var $options = $select.querySelectorAll('option')
+
+  // Create wrapper to inject the autocomplete element
+  $autocompleteWithHintOnOptions.insertAdjacentHTML('beforeend', '<div class="autocomplete__wrapper"></div>')
+
+  // Remove select element from DOM
+  $select.remove()
+
+  new window.accessibleAutocomplete({ // eslint-disable-line no-new, new-cap
+    element: $autocompleteWithHintOnOptions.querySelector('.autocomplete__wrapper'),
+    id: $select.id,
+    source: function (query, syncResults) {
+      var results = []
+      $options.forEach(function ($el) {
+        results.push({value: $el.textContent, hint: $el.dataset.hint || ''})
+      })
+
+      if (query) {
+        results.filter(function (result) {
+          var valueContains = result.value.toLowerCase().indexOf(query.toLowerCase()) !== -1
+          var hintContains = result.hint.toLowerCase().indexOf(query.toLowerCase()) !== -1
+          return valueContains || hintContains
+        })
+      }
+      syncResults(results)
+    },
+    minLength: 3,
+    autoselect: true,
+    showNoOptionsFound: true,
+    templates: {
+      inputValue: function (result) {
+        return result && result.value
+      },
+      suggestion: function (result) {
+        return result && result.value + '<span class="autocomplete__option-hint">' + result.hint + '</span>'
+      }
+    }
+  })
+}
```

Comment:
Should we have this file as a distinctly different JS file since it doesn't share code with the first part?

Reply:
I consider this as not being a different component, but a file where we initialise the autocomplete components – in our case in two different ways.